### PR TITLE
#Task4.1

### PR DESCRIPTION
**Pull request by MEHV29**
This contains the point 4.1 of the thask

### CODE REVIEW

File: Task4/Task4_1/DiagonalMatrix.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+{
+    internal class DiagonalMatrix<T>
+    {
+        private T[] _elementsLocatedOnDiagonal;
+        private int _size;
+
+        public delegate void EventHandler(object sender, ElementChangedEventArgs<T> e);
+
+        public delegate T DelegateMethods(T a, T b);
+
```
Comment (IgorKheidorov) [inline Task4/Task4_1/DiagonalMatrix.cs]: In this case can use generic EventHandler< ElementChangedEventArgs>, not defining it yourself ). Lines 17 and 15 are not needed, remove them.

File: Task4/Task4_1/ElementChangedEventArgs.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Task4_1
+{
```
Comment (IgorKheidorov) [inline Task4/Task4_1/ElementChangedEventArgs.cs:2]: Make it inheriting from EventArgs

File: Task4/Task4_1/DiagonalMatrix.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            {
+                this._elementsLocatedOnDiagonal[i] = elementsLocatedOnDiagonal[i];
+            }
+        }
+
+        protected virtual void OnElementChanged(ElementChangedEventArgs<T> e)
+        {
+            ElementChanged?.Invoke(this, e);
+        }
+
```
Comment (IgorKheidorov) [inline Task4/Task4_1/DiagonalMatrix.cs:87]: Strange method, can' t understand its purpose

File: Task4/Task4_1/ExtensionMethods.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using static Task4_1.Program;
+
+namespace Task4_1
+{
+    internal static class ExtensionMethods
```
Comment (IgorKheidorov) [inline Task4/Task4_1/ExtensionMethods.cs:3]: Better use generic Func<T, T, T> instead of SumDelegate

File: Task4/Task4_1/DiagonalMatrix.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
+﻿using System;
```
Comment (IgorKheidorov) [inline Task4/Task4_1/DiagonalMatrix.cs]: Remove not used usings

**Review by IgorKheidorov — APPROVED** [review]
Not bad) Fix comments and merge

File: Task4/Task4_1/DiagonalMatrix.cs (hunk 1 above)
Comment (MEHV29) [reply Task4/Task4_1/DiagonalMatrix.cs]: Modified

File: Task4/Task4_1/ElementChangedEventArgs.cs (hunk 2 above)
Comment (MEHV29) [reply Task4/Task4_1/ElementChangedEventArgs.cs:2]: Modified

File: Task4/Task4_1/DiagonalMatrix.cs (hunk 3 above)
Comment (MEHV29) [reply Task4/Task4_1/DiagonalMatrix.cs:87]: Added comments to explanation

File: Task4/Task4_1/ExtensionMethods.cs (hunk 4 above)
Comment (MEHV29) [reply Task4/Task4_1/ExtensionMethods.cs:3]: Modified

File: Task4/Task4_1/DiagonalMatrix.cs (hunk 5 above)
Comment (MEHV29) [reply Task4/Task4_1/DiagonalMatrix.cs]: Modified

**Pull request merged**